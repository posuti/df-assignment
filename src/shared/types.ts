export enum ItemStatus {
    pending = "pending",
    ready= "ready",
    error = "error",
    deleted = "deleted"
}

export interface Item {
    id?: string,
    name?: string,
    status?: ItemStatus,
    creationTimeSec?: number,
    deletionTimeSec?: number
}

export enum ApiStatus {
    connecting = "connecting",
    ready= "ready",
    error = "error"
}

export enum SocketStatus {
    connecting = "connecting",
    ready= "connected",
    error = "error"
}