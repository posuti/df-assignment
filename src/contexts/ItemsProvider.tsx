import { useEffect, useState, useReducer, useContext } from 'react';
import { ReactNode, createContext } from "react";
import { ApiStatus, Item } from '../shared/types';
import { useItemsSocket } from '../hooks/useItemsSocketApi';
import ItemsApi from '../services/ItemsApi';

interface Props {
    children: ReactNode;
}

export const ItemsProvider = (props: Props) => {

    const [items, dispatch] = useReducer(itemsReducer, []);
    const [status, setStatus] = useState(ApiStatus.connecting);

    const socketStatus = useItemsSocket(dispatch);

    useEffect(() => {
        setStatus(ApiStatus.connecting);
        let abort = false;
        ItemsApi.GetAllItems().then((newItems: any) => {
            if (abort) return;
            setStatus(ApiStatus.ready);
            dispatch({ type: "inited", payload: newItems });
        }, (err) => setStatus(ApiStatus.error))
        return (() => { abort = true })
    }, []);

    const contextValue = { items, status, socketStatus, dispatch };
    return <ItemsContext.Provider value={contextValue}>{props.children}</ItemsContext.Provider>
}

const ItemsContext = createContext( {} as any );
export const useItemsContext = () => useContext(ItemsContext);

function itemsReducer(state: any, action: any) {
    switch (action.type) {
        case 'inited':
            const items = action.payload || [];
            return [ ...items ];
        case 'added':
            return [...state, { ...action.payload },
            ];
        case 'updated':
            const updatedItem = action.payload;
            const index = state.findIndex((d: Item) => d.id == updatedItem.id);
            if (index > -1) {
                return state.map((item: Item) => item.id == updatedItem.id ? { ...item, status: updatedItem.status } : { ...item });
            } else {
                return [...state, updatedItem];
            }
        default:
            return state;
    }
}