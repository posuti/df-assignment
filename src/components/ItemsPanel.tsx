
import { Alert, Box } from '@mui/material';
import { useItemsContext } from '../contexts/ItemsProvider';
import ItemsTable from './ItemsTable';
import { Item } from '../shared/types';
import { useDeleteItem } from '../hooks/useItemsApi';

function ItemsPanel() {

    const { items, status, socketStatus } = useItemsContext();
    const deleteItem = useDeleteItem();

    function onDelete(itemId: string) {
        deleteItem({
            id: itemId,
            deletionTimeSec: 3
        })
    }

    return (
        <>
            <SoccetStatus status={socketStatus} />
            <ItemsTable status={status} items={items as Item[]} onDelete={(itemId: string) => onDelete(itemId)} />
        </>
    );

}

export default ItemsPanel;

function SoccetStatus({ status }: { status: string }) {

    if (status === "connected") return (
        <Box sx={{ width: '100%' }}>
            <Alert sx={{ mb: 2 }} > Soccet status: connected </Alert>
        </Box>
    );

    return (
        <Box sx={{ width: '100%' }}>
            <Alert sx={{ mb: 2 }} severity="warning">Soccet status: {status}</Alert>
        </Box>
    )
}