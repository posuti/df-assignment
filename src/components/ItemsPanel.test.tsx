
import { useItemsContext } from "../contexts/ItemsProvider";
import { ApiStatus, SocketStatus } from "../shared/types";
import ItemsPanel from "./ItemsPanel";
import renderer from 'react-test-renderer';

jest.mock('../contexts/ItemsProvider', () => ({
    useItemsContext: jest.fn(),
    useDeleteItem: jest.fn(),
}));

describe('ItemsPanel Component', () => {
    it('matches the snapshot', () => {

        const contextParams = {
            items: [{ id: "1" }],
            status: ApiStatus.ready,
            socketStatus: SocketStatus.ready,
            dispatch: jest.fn()
        };

        (useItemsContext as jest.Mock).mockReturnValue(contextParams);

        const tree = renderer.create(<ItemsPanel />).toJSON();
        expect(tree).toMatchSnapshot();

    });
});
