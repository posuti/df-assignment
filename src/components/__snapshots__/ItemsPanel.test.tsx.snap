// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemsPanel Component matches the snapshot 1`] = `
Array [
  <div
    className="MuiBox-root css-8atqhb"
  >
    <div
      className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-standardSuccess MuiAlert-standard css-4bzs8s-MuiPaper-root-MuiAlert-root"
      role="alert"
    >
      <div
        className="MuiAlert-icon css-1ytlwq5-MuiAlert-icon"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
          data-testid="SuccessOutlinedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M20,12A8,8 0 0,1 12,20A8,8 0 0,1 4,12A8,8 0 0,1 12,4C12.76,4 13.5,4.11 14.2, 4.31L15.77,2.74C14.61,2.26 13.34,2 12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0, 0 22,12M7.91,10.08L6.5,11.5L11,16L21,6L19.59,4.58L11,13.17L7.91,10.08Z"
          />
        </svg>
      </div>
      <div
        className="MuiAlert-message css-1pxa9xg-MuiAlert-message"
      >
         Soccet status: connected 
      </div>
    </div>
  </div>,
  <div
    className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root app-panel css-11xur9t-MuiPaper-root-MuiTableContainer-root"
    data-testid="itemsTable"
  >
    <table
      aria-label="simple table"
      className="MuiTable-root css-rqglhn-MuiTable-root"
      role={null}
    >
      <thead
        className="MuiTableHead-root css-15wwp11-MuiTableHead-root"
        role={null}
      >
        <tr
          className="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
          role={null}
        >
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
            scope="col"
            width={300}
          >
            ID
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
            scope="col"
          >
            Name
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
            scope="col"
            width={10}
          >
            Time
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
            scope="col"
            width={70}
          >
            Status
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
            scope="col"
            width={50}
          />
        </tr>
      </thead>
      <tbody
        className="MuiTableBody-root css-apqrd9-MuiTableBody-root"
        role={null}
      >
        <tr
          className="MuiTableRow-root css-34nofg-MuiTableRow-root"
          role={null}
        >
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            scope="row"
          >
            1
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            scope="row"
          >
            Item 1
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            scope="row"
          />
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            scope="row"
          >
            <span
              className="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
              role="progressbar"
              style={
                Object {
                  "height": 20,
                  "width": 20,
                }
              }
            >
              <svg
                className="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
                viewBox="22 22 44 44"
              >
                <circle
                  className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
                  cx={44}
                  cy={44}
                  fill="none"
                  r={20.2}
                  strokeWidth={3.6}
                  style={Object {}}
                />
              </svg>
            </span>
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            scope="row"
          >
            <button
              className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedWarning MuiButton-sizeSmall MuiButton-outlinedSizeSmall Mui-disabled MuiButton-root MuiButton-outlined MuiButton-outlinedWarning MuiButton-sizeSmall MuiButton-outlinedSizeSmall css-5rgykb-MuiButtonBase-root-MuiButton-root"
              disabled={true}
              onBlur={[Function]}
              onClick={[Function]}
              onContextMenu={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={-1}
              type="button"
            >
              Delete
            </button>
          </th>
        </tr>
      </tbody>
    </table>
  </div>,
]
`;
